packages/response-ops/alerts_table/components/alerts_data_grid.tsx

```diff
@@ -0,0 +1,420 @@
+      onColumnResize,
+      showInspectButton = false,
+      leadingControlColumns: additionalLeadingControlColumns,
+      trailingControlColumns,
+      onSortChange,
+      sort: sortingFields,
+      rowHeightsOptions,
+      dynamicRowHeight,
+      alertsQuerySnapshot,
+      additionalToolbarControls,
+      toolbarVisibility: toolbarVisibilityProp,
+      shouldHighlightRow,
+      renderContext,
+      hideBulkActions,
+      casesConfiguration,
+      flyoutAlertIndex,
+      setFlyoutAlertIndex,
+      onPaginateFlyout,
+      onChangePageSize,
+      onChangePageIndex,
+      actionsColumnWidth = DEFAULT_ACTIONS_COLUMN_WIDTH,
+      getBulkActions,
+      fieldsBrowserOptions,
+      cellActionsOptions,
+      pageSizeOptions = DEFAULT_PAGE_SIZE_OPTIONS,
+      height,
+      ...euiDataGridProps
+    } = props;
+    const {
+      isLoading,
+      alerts,
+      alertsCount,
+      isLoadingAlerts,
+      browserFields,
+      renderActionsCell: ActionsCell,
+      pageIndex,
+      pageSize,
+      refresh: refreshQueries,
+      columns,
+      dataGridRef,
+      services: { http, notifications, application, cases: casesService, settings },
+    } = renderContext;
+
+    const { colorMode } = useEuiTheme();
+    const { sortingColumns, onSort } = useSorting(onSortChange, visibleColumns, sortingFields);
+    const {
+      isBulkActionsColumnActive,
+      bulkActionsState,
+      bulkActions,
+      setIsBulkActionsLoading,
+      clearSelection,
+      updateBulkActionsState,
+    } = useBulkActions({
+      ruleTypeIds,
+      query,
+      alertsCount: alerts.length,
+      casesConfig: casesConfiguration,
+      getBulkActions,
+      refresh: refreshQueries,
+      hideBulkActions,
+      http,
+      notifications,
+      application,
+      casesService,
+    });
+
+    const refresh = useCallback(() => {
+      refreshQueries();
+      clearSelection();
+    }, [clearSelection, refreshQueries]);
+
+    const toolbarVisibility = useGetToolbarVisibility({
+      bulkActions,
+      alertsCount,
+      rowSelection: bulkActionsState.rowSelection,
+      alerts,
+      isLoading,
+      columnIds: columns.map((column) => column.id),
+      onToggleColumn,
+      onResetColumns,
+      browserFields,
+      additionalToolbarControls,
+      setIsBulkActionsLoading,
+      clearSelection,
+      refresh,
+      fieldsBrowserOptions,
+      alertsQuerySnapshot,
+      showInspectButton,
+      toolbarVisibilityProp,
+      settings,
+    });
+
+    const customActionsColumn: EuiDataGridControlColumn | undefined = useMemo(() => {
+      if (ActionsCell) {
+        const RowCellRender: EuiDataGridControlColumn['rowCellRender'] = (_props) => {
+          const idx = _props.rowIndex - _props.pageSize * _props.pageIndex;
+          const alert = _props.alerts[idx];
+          const legacyAlert = _props.oldAlertsData[idx];
+          const ecsAlert = _props.ecsAlertsData[idx];
+          const setIsActionLoading = useCallback(
```

Comment:
does this work? a useCallback within a useMemo? Think it will blow up in some cases. Since RowCellRender is just a function that takes props and returns jsx, it should be a component. move it to the top level of this file, wrap in React.memo, should be a little faster, since const RowCellRender is only created once, can be created many times this way.

Reply:
I wanted to convert this to a standalone component but got sidetracked, done! 🙂